[emotiva][WIP] Initial contribution

espenaf opened the pull request:
This PR adds support for Emotiva [AV processors](https://emotiva.com/collections/processors) based on the Emotiva Network Remote Control protocol.

Binding has only tested with an XMC-2, so need feedback from people with other devices. 

Supports both version 2.0 and 3.0 of the protocol.

Basic functionality should be working, but the following functionality is not done or tested:

* Radio tuner
* Zone 2
* Controlling the Menu

mlobstein commented:
The build is failing because your commit it not signed. See here for instructions to fix it: https://github.com/openhab/openhab-addons/pull/16499/checks?check_run_id=22412921872

austvik commented on bundles/org.openhab.binding.anel/src/main/resources/OH-INF/config/config.xml:
Should this file be included in this pull request?

espenaf replied:
Ahh, wrong file. Fixed.

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/java/org/openhab/binding/emotiva/internal/protocol/EmotivaDataType.java:
```diff
@@ -0,0 +1,57 @@
+    FREQUENCY_HERTZ("hertz"),
+    NUMBER("number"),
+    NUMBER_TIME("number_time"),
+    GOODBYE("number_time"),
```
Yet to observe an GOODBYE message from my device, so it is just an assumption. A GOODBYE message will probably just set the device OFFLINE and then we can start an `scheduleConnectRetry(retryConnectInMinutes)` to wait for the device to be available again.

espenaf replied:
Added handling of it.

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,87 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<config-description:config-descriptions
```
Users can set a custom max volume in their device via the setup, but there is no direct API to get that value. However, if you adjust the volume, the min/max values are returned as part of the <emotivaBarNotify> message. It feels a bit to many steps that will actually change status in the device to get this value, so adding a config value probably is the best thing.

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/java/org/openhab/binding/emotiva/internal/EmotivaProcessorHandler.java:
```diff
@@ -0,0 +1,734 @@
+
+                // answer expected within 50-600ms on a regular network; wait up to 2sec just to make sure
+                for (int delay = 0; delay < 10 && !udpSenderActive; delay++) {
+                    Thread.sleep(200); // wait 10 x 200ms = 2sec
```
You are correct, the scheduleConnectRetry will handle any minor or major network issue. This does however give use an simple solution of handling thoese minor network issue without having to wait 2 minutes to a retry. The retry could be reduces to something shorter, but then we anyway probably need a two step retry approach for other issues.

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/java/org/openhab/binding/emotiva/internal/EmotivaProcessorHandler.java:
```diff
@@ -0,0 +1,734 @@
+                        "@text/message.processor.connection.failed");
+                disconnect();
+                scheduleConnectRetry(retryConnectInMinutes);
+            }
```
Not sure, the connection method is synchronized and try to make sure the connection is either working or we need to schedule a retry. Suggestions are welcome.

lsiepel replied:
Regarding the ThingStatus, it is a common practice to:
- In `initialize()`
    - Set `updateStatus(ThingStatus.UNKNOWN);` at the start
    - Check configuration and set `ThingStatus` with details if it is not valid
    - End with a scheduled task (e.g. `updateTask()`) that updates the thing state and channels
- In `updateTask()`
    - Check if connection is authenticated and continue or get authenticated and set `ThingStatus`
    - Get data to update channels, set `ThingStatus` on error.

Not covering all cases and when applied to this binding it might not suit 100%.

espenaf replied:
Updated initialize() with status setting and checking of port and ipaddress to the device.

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,86 @@
+			</description>
+			<advanced>true</advanced>
+		</parameter>
+		<parameter name="retryConnectInMinutes" type="integer" required="false" unit="s">
```
This is not documented in the readme.

lsiepel commented on bundles/org.openhab.binding.emotiva/README.md:
````diff
@@ -0,0 +1,186 @@
+
+`.things` file:
+
+```perl
````
```suggestion
```java
```

espenaf replied:
I lifted this of one of the other bindings, if I change to java, IDEA atleast starts to give errors.

jlaur replied:
In which binding did you find that for a .things definition? This was fixed recently for all bindings - see PR's linked to #13858.

lsiepel commented on bundles/org.openhab.binding.emotiva/README.md:
```diff
@@ -0,0 +1,186 @@
+    }
+    Buttongrid label="Menu Control" staticIcon=material:control_camera item=emotiva_menu_control buttons=[1:1:POWER="Power"=switch-off , 1:2:MENU="Menu", 1:3:INFO="Info" , 2:2:UP="Up"=f7:arrowtriangle_up , 4:2:DOWN="Down"=f7:arrowtriangle_down , 3:1:LEFT="Left"=f7:arrowtriangle_left , 3:3:RIGHT="Right"=f7:arrowtriangle_right , 3:2:ENTER="Select" ]
+}
+...
```
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.emotiva/README.md:
````diff
@@ -0,0 +1,186 @@
+`.sitemap` file:
+
+```perl
+...
````
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.emotiva/README.md:
```diff
@@ -0,0 +1,186 @@
+
+## Full Example
+
+`.things` file:
```
```suggestion
### `.things` file:
```

lsiepel commented on bundles/org.openhab.binding.emotiva/README.md:
````diff
@@ -0,0 +1,186 @@
+Thing emotiva:processor:1 "XMC-2" @ "Living room" [ipAddress="10.0.0.100", protocolVersion="3.0"]
+```
+
+`.items` file:
````
```suggestion
### `.items` file:
```

lsiepel commented on bundles/org.openhab.binding.emotiva/README.md:
````diff
@@ -0,0 +1,186 @@
+String                  emotiva_menu_tottom_end     ""                      <none>  {channel="emotiva:processor:1:general#menu_display_bottom_end"}
+```
+
+`.sitemap` file:
````
```suggestion
### `.sitemap` file:
```

lsiepel commented on bundles/org.openhab.binding.emotiva/README.md:
````diff
@@ -0,0 +1,186 @@
+...
+```
+
+## Network Remote Control protocol Reference
````
```suggestion
## Network Remote Control Protocol Reference
```

lsiepel commented on bundles/org.openhab.binding.emotiva/README.md:
```diff
@@ -0,0 +1,186 @@
+
+## Network Remote Control protocol Reference
+
+These resources can be useful to learn what to send using the `command`channel:
```
```suggestion
These resources can be useful to learn what to send using the `command` channel:
```

lsiepel commented on bundles/org.openhab.binding.emotiva/pom.xml:
```diff
@@ -0,0 +1,25 @@
+
+  <dependencies>
+    <dependency>
+      <groupId>org.assertj</groupId>
```
Is this dependency essential? As i have seen several bindings with unittests without any dependency.

espenaf replied:
I guess its mostly a matter of taste. The whole test suite use the fluent based assertThat() instead of the normal assertEquals/True/False.

lsiepel replied:
Not entirely taste. We try prevent adding unnecessary dependency’s and keep the dependency footprint as small as reasonable possible.

espenaf replied:
I saw that the fineoffsetweatherstation binding was using it, so I though that it would be fine.

lsiepel replied:
Missed that one, in that case let's continue.

jlaur replied:
@espenaf - constaint-based assert model is nice, but perhaps you can use the org.hamcrest matchers instead like many other bindings? See for example: https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.energidataservice/src/test/java/org/openhab/binding/energidataservice/internal/action/EnergiDataServiceActionsTest.java

espenaf replied:
Hehe. One include of assertj might be a mistake, adding one more quickly creates a precedence, so I just rewrote the tests and removed assertj from this binding.

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/java/org/openhab/binding/emotiva/internal/EmotivaConfiguration.java:
```diff
@@ -0,0 +1,41 @@
+ */
+package org.openhab.binding.emotiva.internal;
+
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.DEFAULT_EMOTIVA_PROTOCOL_VERSION;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.DEFAULT_KEEP_ALIVE_IN_MILLISECONDS;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.DEFAULT_RETRY_INTERVAL_MINUTES;
```
```suggestion
import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.*
```

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/java/org/openhab/binding/emotiva/internal/EmotivaUdpBroadcastService.java:
```diff
@@ -0,0 +1,194 @@
+ */
+package org.openhab.binding.emotiva.internal;
+
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.DEFAULT_EMOTIVA_PROTOCOL_VERSION;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.DEFAULT_UDP_SENDING_TIMEOUT;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.THING_PROCESSOR;
```
```suggestion
import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.*
```

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -0,0 +1,11 @@
+	<name>Emotiva Binding</name>
+	<description>This is the binding for devices from the Emotiva Audio Corporation.</description>
+	<connection>local</connection>
+
```
Would be nice if this can be extended with suggestion finder details 
https://www.openhab.org/docs/developer/addons/addon.html#discovery-methods

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/java/org/openhab/binding/emotiva/internal/protocol/EmotivaControlRequest.java:
```diff
@@ -0,0 +1,486 @@
+ */
+package org.openhab.binding.emotiva.internal.protocol;
+
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.CHANNEL_SOURCE;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.CHANNEL_TUNER_BAND;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.CHANNEL_TUNER_CHANNEL_SELECT;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.CHANNEL_ZONE2_SOURCE;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.DEFAULT_TRIM_MAX_DECIBEL;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.DEFAULT_TRIM_MIN_DECIBEL;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.DEFAULT_VOLUME_MAX_DECIBEL;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.DEFAULT_VOLUME_MIN_DECIBEL;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.MAP_SOURCES_MAIN_ZONE;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.MAP_SOURCES_ZONE_2;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.PROTOCOL_V3_LEVEL_MULTIPLIER;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.TRIM_SET_COMMAND_SUFFIX;
+import static org.openhab.binding.emotiva.internal.EmotivaCommandHelper.clamp;
+import static org.openhab.binding.emotiva.internal.EmotivaCommandHelper.volumePercentageToDecibel;
+import static org.openhab.binding.emotiva.internal.protocol.EmotivaCommandType.NUMBER;
+import static org.openhab.binding.emotiva.internal.protocol.EmotivaCommandType.SET;
+import static org.openhab.binding.emotiva.internal.protocol.EmotivaCommandType.TOGGLE;
+import static org.openhab.binding.emotiva.internal.protocol.EmotivaCommandType.UP_DOWN_HALF;
+import static org.openhab.binding.emotiva.internal.protocol.EmotivaCommandType.UP_DOWN_SINGLE;
+import static org.openhab.binding.emotiva.internal.protocol.EmotivaDataType.FREQUENCY_HERTZ;
+import static org.openhab.binding.emotiva.internal.protocol.EmotivaSubscriptionTags.tuner_band;
+import static org.openhab.binding.emotiva.internal.protocol.EmotivaSubscriptionTags.tuner_channel;
```
```suggestion
import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.*
import static org.openhab.binding.emotiva.internal.protocol.*
```

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/java/org/openhab/binding/emotiva/internal/protocol/EmotivaControlRequest.java:
```diff
@@ -0,0 +1,486 @@
+    }
+
+    public EmotivaControlDTO createDTO(Command ohCommand, @Nullable State previousState) {
+
```
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/java/org/openhab/binding/emotiva/internal/protocol/EmotivaControlRequest.java:
```diff
@@ -0,0 +1,486 @@
+            }
+            case UP_DOWN_HALF -> {
+                if (ohCommand instanceof Number value) {
+
```
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/java/org/openhab/binding/emotiva/internal/discovery/EmotivaDiscoveryService.java:
```diff
@@ -0,0 +1,69 @@
+ */
+package org.openhab.binding.emotiva.internal.discovery;
+
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.DISCOVERY_BROADCAST_ADDRESS;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.DISCOVERY_TIMEOUT_SECONDS;
+import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.SUPPORTED_THING_TYPES_UIDS;
```
```suggestion
import static org.openhab.binding.emotiva.internal.EmotivaBindingConstants.*
```

lsiepel reviewed (commented):
Thanks for this contribution. Looked at all except the key controller/handlers, will do in a next round. 
Please check that al SAT errors are fxied and the build comes without any warnings.

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,433 @@
+				<label>Main Zone Control</label>
+				<description>Channels for the main zone of this processor.</description>
+			</channel-group>
+			<channel-group id="zone2" typeId="zone">
```
The API design of the Emotiva Network Remote Control Protocol really ties it to only supporting two zones, a main and zone 2, if they want more zones, they need to extends the set of commands. Looking at the manuals of the possibly supported devices (XMC-1, XMC-2, RMC-1 and RMC-1L), it looks like all of them support this two zones. There are other processors from Emotiva, the BasX series, which do not have zone support, but the do not support this Remote Control Protocol.

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,433 @@
+	<channel-type id="dim">
+		<item-type>Number:Percentage</item-type>
+		<label>Front Panel Dimness</label>
+		<description>Percentage of dimness: "0", "20", "40", "60", "80", "100"</description>
```
The dim command only cycles through these values, so one cannot select a specific value.

lsiepel replied:
Ok, so the UI is covered. Did you technically check if a value like 55 would work from something like a DSL rule ?

espenaf replied:
I was also hoping to find some undocuments features in the protocol when sending other types of values, but no such lock. So if I try to send a exact value on the dim command tag, I just get the following back

`            <emotivaAck>
              <dim status="nak"/>
            </emotivaAck>""";
` 
Which means it did not work.

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,433 @@
+	<channel-type id="tuner_signal">
+		<item-type>String</item-type>
+		<label>Radio Tuner Signal</label>
+		<description>Radio tuner signal quality</description>
```
The value from the processor is a combination of channels and analog signal strength. No real signal will give the value "Mono 0dBuV", while a real signal might give "Stereo 39dBuV". So it really would need two seperate OH channels (audio channels and decibels-microvolt) to be able to use UoM, but as decibels-microvolt (dBuV) is currently not supported, I have just opted for keeping it as a String type.

lsiepel reviewed (changes requested):
^^^

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,433 @@
+		<label>General Control</label>
+		<description>General channels for this processor.</description>
+		<channels>
+			<channel id="power" typeId="mainPower"/>
```
Ahh, my OpenHAB channel name to Emotiva command was working so nice. Anyway got it fixed.

lsiepel replied:
Not fully fixed yet: `typeId="mainPower"` => `typeId="main-power"` (some more )

Now that i look closer it is even better to use the system type: `system.power`, and i think there are more channels that can be changed to pre-defined system types
Docs here: https://www.openhab.org/docs/developer/bindings/thing-xml.html#system-state-channel-types

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,433 @@
+		<item-type>Number:Percentage</item-type>
+		<label>Front Panel Dimness</label>
+		<description>Percentage of dimness: "0", "20", "40", "60", "80", "100"</description>
+		<category>Light</category>
```
Fixed the item-type to Number:Dimensionless and added a state tag now, if I understood `pattern="%d %unit%"` should work?

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,433 @@
+	<channel-type id="selected_movie_music">
+		<item-type>String</item-type>
+		<label>User Selected Mode</label>
+		<description>User-selected movie or music mode for main zone: "Movie" or "Music".</description>
```
Added options

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/i18n/emotiva_no.properties:
```diff
@@ -0,0 +1,146 @@
+# add-on
```
Nice. Will do.

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,86 @@
+			<default>7003</default>
+			<advanced>true</advanced>
+		</parameter>
+		<parameter name="infoPort" type="integer" required="false">
```
Yeah. They are listed EmotivaTransponderDTO response when discovering a device, but according to the documentation this is not being used yet. So I will just remove it for now.

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,433 @@
+	<channel-type id="tuner_program">
+		<item-type>String</item-type>
+		<label>Radio Tuner Program</label>
+		<description>Radio tuner program: "Country", "Rock", "Classical", etc.</description>
```
I have asked Emotiva about it.

espenaf replied:
No word from Emotiva yet, so I have updated it with the offical RDS Program types from https://www.fmsystems-inc.com/rbds-rds-pty-codes-program-types/

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,433 @@
+		<description>Text displayed on front panel bar of device.</description>
+	</channel-type>
+
+	<channel-type id="keep_alive">
```
I have been using it to monitor if the device is available, Emotiva devices do tend to crash sometimes. I have renamed it to last-seen to make it more clear.

lsiepel replied:
Seems like this is more debug related and therefore should not be modelled as channel. What would the use case be for other users?
If the device is unresponsive I guess the thingstqtus should be set to offline with a detailed error message?

espenaf replied:
The org.openhab.binding.emotiva.internal.EmotivaProcessorHandler#checkKeepAliveTimestamp actually does this by setting the device offline. I see your point about the usefullness for others, so I will remove it.

espenaf commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,86 @@
+			<default>7100</default>
+			<advanced>true</advanced>
+		</parameter>
+		<parameter name="model" type="text" required="false" readOnly="true">
```
Move model, revision and dataRevision to property, and populate them during discovery.

lsiepel commented:
While waiting for emotiva to respons to the question about an oiption list, there are some issues to fix.
Can you fix the SAT issues and build warnings. (see
https://github.com/openhab/openhab-addons/actions/runs/9237265388/artifacts/1537689743 )
Some comment about previous review is open: https://github.com/openhab/openhab-addons/pull/16499#discussion_r1613481822

After these are fixed i need one last review round.

espenaf commented:
> While waiting for emotiva to respons to the question about an oiption list, there are some issues to fix. Can you fix the SAT issues and build warnings. (see https://github.com/openhab/openhab-addons/actions/runs/9237265388/artifacts/1537689743 ) Some comment about previous review is open: [#16499 (comment)](https://github.com/openhab/openhab-addons/pull/16499#discussion_r1613481822)
> 
> After these are fixed i need one last review round.

Pushed new changes that fixes all the SAT issues.

lsiepel commented:
> > While waiting for emotiva to respons to the question about an oiption list, there are some issues to fix. Can you fix the SAT issues and build warnings. (see https://github.com/openhab/openhab-addons/actions/runs/9237265388/artifacts/1537689743 ) Some comment about previous review is open: [#16499 (comment)](https://github.com/openhab/openhab-addons/pull/16499#discussion_r1613481822)
> > After these are fixed i need one last review round.
> 
> Pushed new changes that fixes all the SAT issues.

Thanks, Another step forward. There are 35 compile warnings, can you also look into those.

espenaf commented:
> > > While waiting for emotiva to respons to the question about an oiption list, there are some issues to fix. Can you fix the SAT issues and build warnings. (see https://github.com/openhab/openhab-addons/actions/runs/9237265388/artifacts/1537689743 ) Some comment about previous review is open: [#16499 (comment)](https://github.com/openhab/openhab-addons/pull/16499#discussion_r1613481822)
> > > After these are fixed i need one last review round.
> > 
> > 
> > Pushed new changes that fixes all the SAT issues.
> 
> Thanks, Another step forward. There are 35 compile warnings, can you also look into those.

Now those have also been fixed. 

Communication is established with Emotiva, just waiting on word from the people at Engineering.

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/java/org/openhab/binding/emotiva/internal/EmotivaProcessorHandler.java:
```diff
@@ -0,0 +1,787 @@
+            return;
+        }
+
+        if (object instanceof EmotivaAckDTO answerDto) {
```
```suggestion
        if (object instanceof EmotivaAckDTO) {
```

either remove the unused var or (little bit more work but better), override the `EmotivaAckDTO.toString()` and replace `emotivaUdpResponse.answer()` with `answerDto`

espenaf replied:
Added an override on the EmotivaAckDTO.toString()

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/java/org/openhab/binding/emotiva/internal/dto/EmotivaAckDTO.java:
```diff
@@ -0,0 +1,46 @@
+    @XmlAnyElement(lax = true)
+    private List<Object> commands;
+
+    @SuppressWarnings("unused")
+    public EmotivaAckDTO() {
+    }
```
Personal preference i guess, bit this constructor is not used, so why add it and later on supress the warning? 
Just removing would be best. Applies to all other DTO's

espenaf replied:
If there is no empty constructor for each DTO, the `JAXBContext.newInstance(EmotivaAckDTO.class, ...` in EmotivaXmlUtils:75 will fail with the following

```
com.sun.xml.bind.v2.runtime.IllegalAnnotationsException: 1 counts of IllegalAnnotationExceptions

org.openhab.binding.emotiva.internal.dto.EmotivaAckDTO does not have a no-arg default constructor.
	this problem is related to the following location:
		at org.openhab.binding.emotiva.internal.dto.EmotivaAckDTO
			at 
			
com.sun.xml.bind.v2.runtime.IllegalAnnotationsException$Builder.check(IllegalAnnotationsException.java:76)
	at com.sun.xml.bind.v2.runtime.JAXBContextImpl.getTypeInfoSet(JAXBContextImpl.java:450)
	at com.sun.xml.bind.v2.runtime.JAXBContextImpl.<init>(JAXBContextImpl.java:282)
	at com.sun.xml.bind.v2.runtime.JAXBContextImpl.<init>(JAXBContextImpl.java:109)
	at com.sun.xml.bind.v2.runtime.JAXBContextImpl$JAXBContextBuilder.build(JAXBContextImpl.java:1142)
	at com.sun.xml.bind.v2.ContextFactory.createContext(ContextFactory.java:144)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:251)
	at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:240)
	at javax.xml.bind.ContextFinder.find(ContextFinder.java:363)
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:691)
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:632)
	at org.openhab.binding.emotiva.internal.protocol.EmotivaXmlUtils.<init>(EmotivaXmlUtils.java:75)
	at org.openhab.binding.emotiva.internal.AbstractDTOTestBase.<init>(AbstractDTOTestBase.java:28)
	at org.openhab.binding.emotiva.internal.protocol.EmotivaXmlUtilsTest.<init>(EmotivaXmlUtilsTest.java:35)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

```

lsiepel replied:
Ah, this somewhat hidden. Thanks in that case leave the constructors as is.

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,534 @@
+
+	<channel-type id="selected-movie-music">
+		<item-type>String</item-type>
+		<label>User Selected Mode</label>
```
Two channels with the same labels, please change it so that they are distinct.

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,534 @@
+
+	<channel-type id="tuner-channel-select">
+		<item-type>String</item-type>
+		<label>Radio Tuner Channel Name</label>
```
```suggestion
		<label>Tuner Channel Name</label>
```

lsiepel reviewed (changes requested):
Thanks for fixing the warnings and SAT errors. Looks very clean now. These are my final comments.

espenaf commented:
> Thanks for fixing the warnings and SAT errors. Looks very clean now. These are my final comments.

Fixed it all and added a few more translations.

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -260,8 +260,8 @@
 			<options>
-				<option value="preset1">Preset 1</option>
-				<option value="preset2">Preset 2</option>
+				<option value="preset1">@text/channel-type.emotiva.speaker-preset.state.option.preset-1</option>
```
It should not be needed to add a custom key as these options are translated by default. Or what am I missing?

espenaf replied:
Did not work before, so I looked into the Miele binding to find this solution, but I was actually missing the "state" part for every line, so now it works as intended with the short snart in the value.

lsiepel commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -489,7 +489,7 @@
 	<channel-type id="selected-movie-music">
 		<item-type>String</item-type>
-		<label>User Selected Mode</label>
+		<label>User Selected Movie or Music Mode</label>
```
Labels should be 2-3 words and < 23 chars. 
‘Media Mode’ or something similar?

espenaf replied:
Media Mode is a good suggestion.

lsiepel commented:
Now, you could add your binding's logo to the openHAB website. `
See https://www.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website

Please also see the updated documentation here - SVG is now preferred:
https://next.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website-and-the-ui

espenaf commented:
> Now, you could add your binding's logo to the openHAB website. ` See https://www.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website
> 
> Please also see the updated documentation here - SVG is now preferred: https://next.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website-and-the-ui

Thanks for the good work in reviewing it. I always like this attention to detail.

Opened https://github.com/openhab/openhab-docs/pull/2310 for the logo in SVG format.

jlaur commented on bundles/org.openhab.binding.emotiva/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -0,0 +1,31 @@
+	<connection>local</connection>
+
+	<discovery-methods>
+		<discovery-method>
+			<service-type>ip</service-type>
+			<discovery-parameters>
+				<discovery-parameter>
+					<name>type</name>
+					<value>ipBroadcast</value>
+				</discovery-parameter>
+				<discovery-parameter>
+					<name>destPort</name>
+					<value>7001</value>
+				</discovery-parameter>
+				<discovery-parameter>
+					<name>timeoutMs</name>
+					<value>1000</value>
+				</discovery-parameter>
+			</discovery-parameters>
+		</discovery-method>
```
@espenaf - how did you test this? I just noticed this started appearing in my logs:

```
2024-06-18 21:54:43.493 [WARN ] [fig.discovery.addon.ip.IpAddonFinder] - binding-emotiva: discovery-parameter 'request' or 'requestPlain' required
```

This discovery method seems incomplete as it also lacks `match-properties`. For a full example, see here:

https://github.com/openhab/openhab-addons/blob/54ba2c732b0f12f2df5d6de72723fcd018de73f9/bundles/org.openhab.binding.danfossairunit/src/main/resources/OH-INF/addon/addon.xml#L11-L39

espenaf replied:
I did test the discovery, so that is working. Does this missing request or requestPlain parameter block the actual discovery in any way, or is it just helpful metadata?

The reason I did not add any request property is that the discovery needs this EmotivaPing XML structure, so it did not thing it would fit for this values. I see now that the ipcamera binding is having the same "problem", and it has actually added a XML escaped value in the requestPlain, so I could probably do the same. Something like:

```
				<discovery-parameter>
					<name>requestPlain</name>
					<value>&lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-8&quot;?&gt;&lt;emotivaPing protocol=&quot;2.0&quot;/&gt;</value>
				</discovery-parameter>
```

jlaur replied:
Just to be sure, when you mention you tested the discovery, do you mean the suggested binding feature:

![image](https://github.com/openhab/openhab-addons/assets/19519842/c652fc2c-936a-4844-a643-43ebd88b058d)

or the discovery within the binding?

The logged message means there's an error in the defined `discovery-method`, and it will not work.

espenaf replied:
Ohh, the suggested binding feature was news to me. Then the whole discovery-methods made a bit more sense. After some tweaking of the values, for instance changing ipBroadcast to ipMulticast and adding requestPlain and adding match-properties with reponse=.*, it has stopped giving warnings, but it is not actually discovering anything yet. 

How do I trigger the discovery to start? I got DEBUG setup for org.openhab.core.config.discovery.addon so I am seeing what bundles are trying to do discovery, but my binding is not being triggered. I am running this on openhab 4.1.3 with the newly compiled code added as a jar file.

lsiepel replied:
Please create a PR with your changes and we can continue discussion from there.

espenaf replied:
Opened https://github.com/openhab/openhab-addons/pull/16891
